Kubernetes/HOWTO-on-prem.md

```diff
@@ -102,91 +75,219 @@ sudo cp -r !(*.tgz) ${CNI_BIN}
-
-    ~/kube $ generate-routes.sh 192.168
 
 ### Prepare Manifests & Addons ###
 In the `manifest` folder, run the Python script, passing your master IP and the _full_ cluster CIDR:
 
-    $ python2 generate.py 10.123.45.67 192.168.0.0/16
+    $ ./generate.py $MASTER_IP 192.168.0.0/16
+
+This will generate a set of YAML files. You should [re]move the Python script so that Kubernetes doesn't mistake it for a manifest; this will cause problems.
+
+
+### *Optional*: Customize Cluster CIDR ###
+It's possible that you don't have a continous /16 open for cluster routing. This requires more effort on your part as the configurator, as you'll have to explicitly specify per-node CIDRs that you have available. First, you will need to modify the `kube-controller-manager.yaml` configuration file that we just generated. Search for this line:
+
+    - "--allocate-node-cidrs=True"
+
+and change it to `=False`. Now, Kubernetes will not automatically assign subnets to new nodes. Read [this section](#optional-customize-windows-node-cidr) to know what to do on the Windows nodes; the same logic applies to the `start-kubelet.sh` script on the master, and any Linux worker node `kubelet` calls.
 
-This will generate a set of YAML files. You should [re]move the Python script so that K8s doesn't mistake it for a manifest and cause problems.
 
 ### Configure & Run Kubernetes ###
-Run the script `configure-kubectl.sh` passing your external IP again. This will create a configuration in `~/.kube/config` containing all of the certificate data and other parameters:
+Configure Kubernetes with the certificates we generated previously. This will create a configuration at `~/.kube/config`:
+
+    ~/kube $ ./configure-kubectl.sh $MASTER_IP
 
-    ~/kube $ ./configure-kubectl.sh 10.123.45.67
+Finally, we're ready to start `kubelet`, the Kubernetes "client." This script runs indefinitely, so open another shell afterward to keep working:
 
-Run the script `start-kubelet.sh`. You may need to run it with `sudo`. Then, in another shell, watch your working directory (`~/kube`) until the folder `kubelet/` appears. Immediately copy the Kubernetes configuration file to it:
+    ~/kube $ sudo ./start-kubelet.sh
+
+Watch your working directory until the `kubelet` _folder_ appears in it. Immediately copy the Kubernetes configuration file to it:
 
     ~/kube $ sudo cp ~/.kube/config kubelet/
 
-**TODO**: Why is this necessary? Shouldn't this be done automatically by one of the containers?
+You may need to restart the `start-kubelet.sh` script after this if it wasn't picked up fast enough.
+
+> This is a quirky workaround because for whatever reason, the API server pod doesn't mount the configuration file like it should.
+
+In yet another terminal session, run the Kubeproxy script, passing your cluster CIDR
 
-In another terminal session, run the Kubeproxy script, passing your cluster CIDR (you may also need to run this one with `sudo`):
+    ~/kube $ sudo ./start-kubeproxy.sh 192.168

```

Comment:
Does it matter if we're using upstream routed mode with static POD CIDRs specified on each host if the cluster CIDR referenced isn't a full /16? Can it be any prefix with mask that matches the POD CIDR range across the cluster?

Reply:
Can we start both kubelet and kube-proxy referenced in these scripts (start-kubeproxy.sh and start-kubelet.sh) in the background using "&" and then redirect stdout and stderr to some log file?

Reply:
Definitely. This is something users can just do on their own as they wish (they can even make them a systemd service that runs on startup, etc.) but I'll add a note.

Reply:
As for Kubeproxy, that's actually a really good question. I'm not sure. The [documentation](https://kubernetes.io/docs/admin/kube-proxy/) says:

 > `--cluster-cidr string`
>
> The CIDR range of pods in the cluster. When configured, traffic sent to a Service cluster IP from outside this range will be masqueraded and traffic sent from pods to an external LoadBalancer IP will be directed to the respective cluster IP instead

This implies to me that if you have a weird cluster configuration, where, say, master is on 192.168.0.0/28 and node 1 is on 192.168.5.0/31, you would *still* have to specify 192.168/16. **But** because it only applies to traffic going to service IPs and load balancer IPs, it wouldn't interfere with traffic on other subnets that isn't destined for these. 

Really good catch; I'll make note of that in this section.